File: src/main/java/com/morak/performancetracker/aop/persistence/QueryMonitorAop.java
package com.morak.performancetracker.aop.persistence;

import com.morak.performancetracker.aop.util.ProxyFactoryBeanUtils;
import com.morak.performancetracker.context.Accumulator;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.springframework.stereotype.Component;

@Component
@Aspect
public class QueryMonitorAop {

    private final QueryMonitor queryMonitor;
    private final Accumulator accumulator;

    public QueryMonitorAop(QueryMonitor queryMonitor, Accumulator accumulator) {
        this.queryMonitor = queryMonitor;
        this.accumulator = accumulator;
    }

    @Around("execution(* javax.sql.DataSource.getConnection())")
    public Object datasource(ProceedingJoinPoint proceedingJoinPoint) throws Throwable {
        Object returnValue = proceedingJoinPoint.proceed();
        ConnectionAdvisor advisor = new ConnectionAdvisor(new ConnectionPointcut(),
                new ConnectionAdvice(queryMonitor, accumulator));
        return ProxyFactoryBeanUtils.createObject(returnValue, false, advisor);
    }
}

File: src/main/kotlin/com/morak/performancetracker/aop/persistence/QueryMonitorAop.kt
package com.morak.performancetracker.aop.persistence

import com.morak.performancetracker.aop.util.ProxyFactoryBeanUtils
import com.morak.performancetracker.context.Accumulator
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.springframework.stereotype.Component

@Component
@Aspect
class QueryMonitorAop(private val queryMonitor: QueryMonitor, private val accumulator: Accumulator) {
    @Around("execution(* javax.sql.DataSource.getConnection())")
    fun datasource(proceedingJoinPoint: ProceedingJoinPoint): Any? {
        val returnValue = proceedingJoinPoint.proceed()
        val advisor = ConnectionAdvisor(
            ConnectionPointcut(),
            ConnectionAdvice(queryMonitor, accumulator)
        )
        return ProxyFactoryBeanUtils.createObject(returnValue, false, advisor)
    }
}
